### PR TITLE
Ability to SCP download a file or directory from instance to host

### PR DESCRIPTION
**Pull request by ytsarev**
Closest use case is to get junit compatible report from test suite
and feed it to Jenkins

### CODE REVIEW

**Comment by ytsarev** [comment]
@fnichol  @tyler-ball Review anybody ?

File: lib/kitchen/driver/ssh_base.rb [hunk 1]
```diff
@@ -115,6 +115,14 @@ def verify(state) # rubocop:disable Metrics/AbcSize
       rescue Kitchen::Transport::TransportFailed => ex
         raise ActionFailed, ex.message
       ensure
+        if verifier[:scp_file]
```
Comment (tyler-ball) [inline lib/kitchen/driver/ssh_base.rb]: Why did you decide to put this in the `ensure` block?  I would think you would want it to go in the existing connection block.

Comment (tyler-ball) [inline lib/kitchen/driver/ssh_base.rb]: There should also be some debug logging after the download finishes

**Comment by tyler-ball** [comment]
I think `download` and its specs look good.  Can you update https://github.com/test-kitchen/test-kitchen/blob/master/spec/kitchen/driver/ssh_base_spec.rb to add coverage to the changes you made in `ssh_base.rb`?

File: lib/kitchen/driver/ssh_base.rb (hunk 1 above)
Comment (ytsarev) [inline lib/kitchen/driver/ssh_base.rb]: In fact I want to download junit.xml with detailed report of serverspec failures.

**Comment by ytsarev** [comment]
@tyler-ball i've addressed the issues you pointed out, could you review again ?

File: lib/kitchen/driver/ssh_base.rb (hunk 1 above)
Comment (tyler-ball) [inline lib/kitchen/driver/ssh_base.rb]: I still don't think this needs to be in an `ensure` block.  If your serverspec command fails, does it raise an exception?  If so, which one?

There are already 4 commands being run on the remote host in the `begin`/`rescue` block.  If any of those 4 fail except the serverspec command then the `download` command below will also fail.  That could mask a real failure by raising, instead, a failure from the `download` command.

Does my concern make sense?

**Comment by tyler-ball** [comment]
@ytsarev Thanks for the update specs.  I only had 1 other comment that I will wait for your feedback on

File: lib/kitchen/driver/ssh_base.rb (hunk 1 above)
Comment (ytsarev) [inline lib/kitchen/driver/ssh_base.rb]: Concern are very valid.

However I'm not sure how to proceed. My use case is the use of shell verifier( https://github.com/test-kitchen/test-kitchen/pull/741) which runs wrapper around serverspec test suite. 

If the wrapper returns non-zero exit status it raises ssh TransportFailed ...
This behavior aslo relates to ssh transport design.

So in fact I'm not sure how to proceed with moving it out of `ensure` and avoid shotgun surgery of ssh transport.

Do you have any idea ?

Comment (tyler-ball) [inline lib/kitchen/driver/ssh_base.rb]: Paging @fnichol to get his opinion

Comment (ytsarev) [inline lib/kitchen/driver/ssh_base.rb]: @fnichol I will be grateful for review here

**Comment by cliles** [comment]
So is this good to merge? +1 on the use case, just what I am trying to do.

**Comment by ytsarev** [comment]
@tyler-ball @fnichol please decide on this one

**Comment by tas50** [comment]
@fnichol This would be hugely useful in some work I'm noodling on with automated Ohai testing and I know @lamont-granquist had mentioned the need for this same sort of feature for omnibus work.

**Comment by thommay** [comment]
supercedes #130 i guess :) be good to see this fixed.

**Comment by yyuu** [comment]
:+1:

Now I'm doing dirty `instance_eval` to do this kind of trick....

**Comment by ytsarev** [comment]
@tyler-ball @fnichol so Christmas present anybody ?

**Comment by cheeseplus** [comment]
So we totally want to get this one in and given that WinRM needs more work we don't have to gate on that. That said could we get two things?

1)  raise NotImplementedError in the base transport to if you are on WinRM
2)  have a command, ex `kitchen download foo.txt` wired up

**Comment by ytsarev** [comment]
@cheeseplus alright, thanks, sounds like a plan. Will ping you back when implementation ready.

**Comment by cheeseplus** [comment]
@ytsarev awesome and if you need any help or get stuck feel free to holler

**Comment by ytsarev** [comment]
@cheeseplus Sorry, it's been a while, finally returned to this one
* I've implemented 1), please review. 
* I believe that 2) here is irrelevant and should be implemented as separate additional feature as we have no symmetric upload command
What do you think ?

**Comment by ytsarev** [comment]
@cheeseplus ping

**Comment by KierranM** [comment]
Given that `winrm-fs` already has the functionality to download files it would be great to add this to winrm transport as well, I have essentially the same use case as @ytsarev of getting Pester results into TeamCity.

From what I can gather it would only involve adding:
```ruby
# (see Base::Connection#download)
def download(remote, local)
  file_transporter.download(remote, local)
end
```
to the WinRM transport.

Is there a reason, @cheeseplus, that you believe it would require enough work to not be included?

**Comment by mwrock** [comment]
adding winrm compatibility should be trivial but it could certainly be added in a separate PR. I'd hate to see it block this. From what I can tell the only open item here appears to be implementation details around the `ensure` block. Those concerns do look valid.

My biggest concern there is that if an error occurs in one of the previous commands and then the ensure block also fails, we lose the original error.

**Comment by mwrock** [comment]
Given we want download to be transport agnostic, we should use different setting names that don't contain `scp`

File: lib/kitchen/driver/ssh_base.rb (hunk 1 above)
Comment (AgarFu) [inline lib/kitchen/driver/ssh_base.rb]: I'd love tho have this feature merged, looks like the the discussion about how to solve the rescue thing is not going anywhere. I do understand the concerns there is no an easy way to solve this issue.

Maybe to add as much as detail to the ActionFailed and SshFailed exceptions as possible to help the user understand the reason why the process failed is the only solution without change the ssh transport design.

IMHO this feature shouldn't be blocked any more time by this issue, I think this is super important to help identify errors in executions of more tan 3 or 4 tests suites in parallel. With the current implementation try to find the error in the log when you're running 15 or 16 parallel suites is a real pain.

👍

**Comment by cheeseplus** [comment]
Closing in favor of https://github.com/test-kitchen/test-kitchen/pull/1306

**Pull request closed without merging**